src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,140 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
     @Override
     public void add(T value) {
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Wrong index");
+        }
+        if (index == size) {
+            Node<T> lastNode = last;
+            Node<T> newNode = new Node<>(lastNode, value, null);
+            if (first == null) {
+                first = newNode;
+            } else {
+                last.next = newNode;
+            }
+            last = newNode;
+            size++;
+        } else if (index > 0) {
+            Node<T> newNode = first;
+            for (int i = 0; i < index; i++) {
+                newNode = newNode.next;
+            }
+            Node<T> currentNode = new Node<>(newNode.prev, value, newNode);
+            currentNode.prev.next = currentNode;
+            currentNode.next.prev = currentNode;
+            size++;
+        } else {
+            Node<T> newNode = first;
+            Node<T> currentNode = new Node<>(null, value, newNode);
+            newNode.prev = currentNode;
+            first = currentNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> newNode = findNodeByIndex(index);
+        return newNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> newNode = findNodeByIndex(index);
+        T oldValue = newNode.value;
+        newNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> newNode = findNodeByIndex(index);
+        if (first == last) {
+            first = null;
+            last = null;
+        } else if (newNode == first) {
+            newNode.next.prev = newNode.prev;
+            first = newNode.next;
+        } else if (newNode == last) {
+            newNode.prev.next = newNode.next;
+            last = newNode.prev;
+        } else {
+            newNode.prev.next = newNode.next;
+            newNode.next.prev = newNode.prev;
+        }
+        size--;
+        return newNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> newNode = first;
+        int index = 0;
+        while (newNode != null) {
+            if (newNode.value == null && object == null || newNode.value.equals(object)) {
+                remove(index);
```

Comment:
![image](https://user-images.githubusercontent.com/81781479/183491214-9075c3ce-fd4a-442c-90fe-3ff4708bb547.png)
when you call remove, it iterates over linked list again, at this moment you can just unlink this node